file/hasher/hasher_test.go

```diff
@@ -0,0 +1,179 @@
+						bmtobj := bmt.New(pool)
+						d := data[:n]
+						rbmtobj := bmt.NewRefHasher(hasher, count)
+						expNoMeta := rbmtobj.Hash(d)
+						h := hasher()
+						h.Write(bmt.ZeroSpan)
+						h.Write(expNoMeta)
+						exp := h.Sum(nil)
+						ctx, cancel := context.WithCancel(context.Background())
+						defer cancel()
+						sw := NewAsyncHasher(ctx, bmtobj, double, nil)
+						got := asyncHashRandom(sw, 0, d, wh)
+						if !bytes.Equal(got, exp) {
+							t.Fatalf("wrong async hash (asyncpart) for datalength %v: expected %x, got %x", n, exp, got)
+						}
+					}
+				})
+			}
+		}
+	}
+}
+
+func BenchmarkBMTAsync(t *testing.B) {
+	whs := []whenHash{first, last, random}
+	for size := 4096; size >= 128; size /= 2 {
+		for _, wh := range whs {
+			for _, double := range []bool{false, true} {
+				t.Run(fmt.Sprintf("double_%v_hash_when_%v_size_%v", double, wh, size), func(t *testing.B) {
+					benchmarkBMTAsync(t, size, wh, double)
+				})
+			}
+		}
+	}
+}
+
+// splits the input data performs a random shuffle to mock async section writes
+func asyncHashRandom(bmtobj *AsyncHasher, spanLength int, data []byte, wh whenHash) (s []byte) {
+	idxs, segments := splitAndShuffle(bmtobj.SectionSize(), data)
+	return asyncHash(bmtobj, spanLength, len(data), wh, idxs, segments)
+}
+
+// mock for async section writes for file.SectionWriter
+// requires a permutation (a random shuffle) of list of all indexes of segments
+// and writes them in order to the appropriate section
+// the Sum function is called according to the wh parameter (first, last, random [relative to segment writes])
+func asyncHash(bmtobj *AsyncHasher, spanLength int, l int, wh whenHash, idxs []int, segments [][]byte) (s []byte) {
+	bmtobj.Reset()
+	if l == 0 {
+		bmtobj.SetSpan(spanLength)
+		return bmtobj.SumIndexed(nil, l)
+	}
+	c := make(chan []byte, 1)
+	hashf := func() {
+		bmtobj.SetSpan(spanLength)
+		c <- bmtobj.SumIndexed(nil, l)
+	}
+	maxsize := len(idxs)
+	var r int
+	if wh == random {
+		r = rand.Intn(maxsize)
+	}
+	for i, idx := range idxs {
+		bmtobj.WriteIndexed(idx, segments[idx])
+		if (wh == first || wh == random) && i == r {
+			go hashf()
+		}
+	}
+	if wh == last {
+		bmtobj.SetSpan(spanLength)
+		return bmtobj.SumIndexed(nil, l)
+	}
+	return <-c
+}
+
+// benchmarks BMT hasher with asynchronous concurrent segment/section writes
+func benchmarkBMTAsync(t *testing.B, n int, wh whenHash, double bool) {
+	data := testutil.RandomBytes(1, n)
+	hasher := sha3.NewLegacyKeccak256
+	pool := bmt.NewTreePool(hasher, bmttestutil.SegmentCount, bmt.PoolSize)
+	bmth := bmt.New(pool)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	bmtobj := NewAsyncHasher(ctx, bmth, double, nil)
+	idxs, segments := splitAndShuffle(bmtobj.SectionSize(), data)
+	rand.Shuffle(len(idxs), func(i int, j int) {
+		idxs[i], idxs[j] = idxs[j], idxs[i]
+	})
+
+	t.ReportAllocs()
+	t.ResetTimer()
+	for i := 0; i < t.N; i++ {
+		asyncHash(bmtobj, 0, n, wh, idxs, segments)
```

Comment:
Could this benchmark be unexpectedly optimised by compiler as the function result is not used anywhere?